design/026-requirements_traits.md

```diff
@@ -0,0 +1,125 @@
+- ``build`` (default False): This dependency is a build tool, an application or executable, like cmake, that is used exclusively at build time, it is not linked/embedded into binaries, and will be in the build context.
+- ``run`` (default None): This dependency contains some executables, either apps or shared libraries that need to be available to execute (typically in the path, or other system env-vars). By default is None, indicating that Conan can try to deduce if the dependency is needed to execute (if ``options.shared`` is True). This trait can be True for ``build=False``, in that case, the package will contain some executables that can run in the host system when installing it, typically like an end-user application. This trait can be True for ``build=True``, the package will contain executables that will run in the build context, typically while being used to build other packages.
+- ``visible`` (default True): This ``require`` will be propagated downstream, even if it doesn’t propagate ``headers``, ``libs`` or ``run`` traits. Requirements that propagate downstream can cause version conflicts. This is by default True, because in most cases, having 2 different versions of the same library in the same dependency graph is at least complicated, if not directly violating ODR or causing linking errors.
+- ``transitive_headers`` (default None): The headers of the dependency will be visible downstream or not. The default None allows Conan to auto detect this trait, for example, if the current package is a ``header-only`` one, and it depends on another library (header only, or static/shared), the headers of the transitive dependency must be available and used in the ``-I<includedirs>`` compilation downstream.
+- ``transitive_libs`` (default None): The libraries to link with of the dependency will be visible downstream or not. The default None allows Conan to auto detect this trait, for example, if the current package is a ``header-only`` one, and it depends on another library (header only, or static/shared), the libraries of the transitive dependency must be available and used in the ``-I<libs>`` and ``-L<libdirs>`` compilation downstream.
+- ``test`` (default False): this requirement is a test library or framework, like Catch2 or gtest. It is mostly a library that needs to be included and linked, but that will not be propagated downstream.
+- ``package_id`` (default None): if the recipe wants to specify how the dependency version affects the current package ``package_id``, can be directly specified here. While it could be also done in the ``package_id()`` method, it seems simpler to be able to specify it in the ``requires`` while avoiding some ambiguities.
+- ``force`` (default False): This ``requires`` will force its version in the dependency graph upstream, overriding other existing versions even of transitive dependencies, and also solving potential existing conflicts.
```

Comment:
What happens if a package downstream forces a different version of already forced dependency?

For example, for graph A <- B <- C <- D, lets's say that `C` forces the dependency on `A/1.0.0` and `D` forces the dependency on `A/2.0.0`. Do we get a conflict or the final resolved version will be `A/2.0.0` (most-downstream wins, as today)?

Reply:
The more downstream always win, I think that is a good principle that works well in Conan, let the user have the power and control. If your recipe declares a "force" on A/2.0.0, that will be the dependency used, no matter what the other upstream dependencies do.

Actually, the explicit definition of the ``force`` trait, allows a much better behavior, allowing not accidentally overriding from downstream, detecting conflicts better, and only forcing overrides when intended (replacing the relatively ugly CONAN_ERROR_ON_OVERRIDE current env-var).

Reply:
OK, I still don't get it. 

If C's recipe declares "force" on A/1.0.0 and D's recipe declares "force" on A/2.0.0, will C's "force" prevent D to force A/2.0.0 or D's desire will be fulfilled and C's "force" will be ignored (in that case, C's `force` doesn't actually do anything...)?

Reply:
I am also not sure I understand this trait. I feel it could lead to some very hard to understand, and impossible to resolve conflicts based on what was already said above

Reply:
D ``force`` desire will always be fullfilled (same as today). This is the only way to make sure it is always possible to resolve all conflicts. 

There is nothing inherently bad to that, in the same way a regular require can be change/overriden from downstream, a force is nothing more than an indication that such requirement should prevail. If we follow the rule that downstream consumers always rule, then the force of D has higher precedence than the force of C. There is no ambiguity, no problem of not resolving conflicts. D forced A/2.0, so A/2.0 will be used upstream, no possibility of conflicts.

This is exactly how the Conan current overrides model works right now, just made explicit (and changing the default conflicting policy).

Reply:
@memsharded, but in that case, I don't see the difference between `force=True` and `force=False` - according to your explanation in both cases, the final version of A used will be `A/2.0`.

Or maybe in the case when C forces `A/1.0.0` and D requires (without force) `A/2.0.0`, the final selected version of A would be `1.0.0`?

Reply:
Only if D declare A/2.0.0 as forced, that will be forced upstream. If not declared as either "forced" or "override" it will conflict if the version doesn't match (behavior CONAN_ERROR_ON_OVERRIDE, introduced a few releases ago), which would be the case if C declared A/1.0 as forced.

That allows all possible desired behaviors: conflict, pure override (no direct dependency), dependency forced (conflicts solved)

Reply:
I like the conceptual model, but I don't love the wording, since it allows saying nonsensical things like `self.requires(...,forced=True, override=False)` (is that just a normal requires?). Or questionable things like `self.requires("foo/[^1.2]",override=True)` (it seems odd to override with a non-specific version range).

I also feel like spelling this in `self.requires(...,override=True)` might be problematic regarding requests like https://github.com/conan-io/conan/issues/2140, https://github.com/conan-io/conan/issues/8017 that have asked to allow overriding via the context (command-line or profile). Presumably one can't, or at least shouldn't, add arbitrary new requires via the command-line or a profile, but if it's all done via calling `self.requires` with various kwargs the boundary of which flags you could use is a bit blurry.

Maybe separate it into two methods (both to be called inside self.requirements() - self.requires(...) to establish a direct requirement and the propagation kwargs, and `self.override("name/version@user/channel")` just to specify that a particular reference is overridden if required(in both self and upstream)? To get the effect of `forced=True` (use it directly *and* override upstream packages), just call both. Or one could allow just `self.override("name")` when you've already got a requires that picked a specific version in this recipe. But maybe not - overrides are exotic enough I'm not sure they need that much syntax sugar.

Since that gets it down to a single string-value argument, it would also lend itself to a declarative `self.overrides = "name1/version@user/channel","name2"` attribute form, and thus to something that could be fit a CLI `--overrides` argument or profile `[overrides]` section...

Reply:
So if I explicitly call out a dependency, let's say `openssl/1.1.1k` (because I want to make sure that is the version used), if somewhere in the dependency graph there is a requires `openssl/1.1.1f` with `force=True` that would override the newer version I explicitly specified?

Reply:
I suggest separating conflict resolution with defining dependencies into separate functions. Their interface is orthogonal and can't (shouldn't ?) be mixed: If I want to resolve conflict (or put some constraint on package version) other parameters (like `visible` or `headers`) are ignored (or at least should - it doesn't make much sense to affect how package is consumed by other packages, might be even impossible).

In other words provide a separate function with clear intent for that, like `constraint_requirement` or so.

Reply:
> So if I explicitly call out a dependency, let's say `openssl/1.1.1k` (because I want to make sure that is the version used), if somewhere in the dependency graph there is a requires `openssl/1.1.1f` with `force=True` that would override the newer version I explicitly specified?

Still wondering if a specifically called out dependency gets overridden by a `force` upstream?  Seems like it could be hard to trace _why_ were getting a package version that may not be what we intended.